language/compiler/bytecode_source_map/src/source_map.rs

```diff
@@ -0,0 +1,319 @@
+    pub type_parameters: Vec<SourceName>,
+
+    // The index into the vector is the locals index. The corresponding `(Identifier, Loc)` tuple
+    // is the name and location of the local.
+    pub locals: Vec<SourceName>,
+
+    // The source location map for the function body.
+    pub code_map: BTreeMap<CodeOffset, Loc>,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct ModuleSourceMap {
+    // The name <address.module_name> for module that this source map is for
+    pub module_name: (AccountAddress, Identifier),
+
+    // A mapping of StructDefinitionIndex to source map for each struct/resource
+    struct_map: HashMap<TableIndex, StructSourceMap>,
+
+    // A mapping of FunctionDefinitionIndex to the soure map for that function.
+    function_map: HashMap<TableIndex, FunctionSourceMap>,
+}
+
+impl StructSourceMap {
+    pub fn new(decl_location: Loc) -> Self {
+        Self {
+            decl_location,
+            type_parameters: Vec::new(),
+            fields: Vec::new(),
+        }
+    }
+
+    pub fn add_type_parameter(&mut self, type_name: SourceName) {
+        self.type_parameters.push(type_name)
+    }
+
+    pub fn get_type_parameter_name(&self, type_parameter_idx: usize) -> Option<SourceName> {
+        self.type_parameters.get(type_parameter_idx).cloned()
+    }
+
+    pub fn add_field_location(&mut self, field_loc: Loc) {
+        self.fields.push(field_loc)
+    }
+
+    pub fn get_field_location(&self, field_index: FieldDefinitionIndex) -> Option<Loc> {
+        self.fields.get(field_index.into_index()).cloned()
+    }
+}
+
+impl FunctionSourceMap {
+    pub fn new(decl_location: Loc) -> Self {
+        Self {
+            decl_location,
+            type_parameters: Vec::new(),
+            locals: Vec::new(),
+            code_map: BTreeMap::new(),
+        }
+    }
+
+    pub fn add_type_parameter(&mut self, type_name: SourceName) {
+        self.type_parameters.push(type_name)
+    }
+
+    pub fn get_type_parameter_name(&self, type_parameter_idx: usize) -> Option<SourceName> {
+        self.type_parameters.get(type_parameter_idx).cloned()
+    }
+
+    /// A single source-level instruction may possibly map to a number of bytecode instructions. In
+    /// order to not store a location for each instruction, we instead use a BTreeMap to represent
+    /// a segement map (holding the left-hand-sands of the segments).  Thus, an instruction
+    /// sequence is always marked from the starting point. To determine what part of the source
+    /// code corresponds to a given `CodeOffset` we query to find the element that is the largest
+    /// number less than or equal to the query. This will give us the location for that bytcode
+    /// range.
+    pub fn add_code_mapping(&mut self, start_offset: CodeOffset, location: Loc) {
+        self.code_map.insert(start_offset, location);
+    }
+
+    // NB that locations must be added in order.
```

Comment:
What is NB?

Reply:
"Nota Bene" or note well (https://en.wikipedia.org/wiki/Nota_bene). I'll go through and update with something more readable.